File: Data Structures & Algorithms/valid-binary-search-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    # keep track of parent node as well as comparing if left and right 
    # is valid

    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def validRec(root, leftBound, rightBound):
            if root == None:
                return True
            
            if root.val <= leftBound or root.val >= rightBound:
                return False
            
            return validRec(root.left, leftBound, root.val) and validRec(root.right, root.val, rightBound)
            
        
        return validRec(root, float("-inf"), float("inf"))
        
        

File: Data Structures & Algorithms/insert-into-a-binary-search-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void dfs(TreeNode* root, int val) {
        if (!root) {
            return;
        }

        if (val > root->val) {
            if (root->right) {
                dfs(root->right, val);
            } else {
                root->right = new TreeNode(val);
            }
        }
        else {
            if (root->left) {
                dfs(root->left, val);
            } else {
                root->left = new TreeNode(val);
            }
        }
    }

    TreeNode* insertIntoBST(TreeNode* root, int val) {
        if (!root) {
            return new TreeNode(val);
        }

        dfs(root, val);
        return root;
    }
};